src/drivers/fluid_wasapi.c

```diff
@@ -0,0 +1,811 @@
+ *    to use the default Windows audio device for media playback.
+ * 
+ * Notes:
+ *  - If exclusive mode is selected, audio.period-size is used as the periodicity
+ *    of the IAudioClient stream, which is the sole factor of audio latency.
+ *    audio.periods still determines the buffer size, but has no direct impact on
+ *    the latency (at least according to Microsoft). The valid range for
+ *    audio.period-size may vary depending on the driver and sample rate.
+ *  - The sample rate and sample format of fluidsynth must be supported by the
+ *    audio device in exclusive mode. Otherwise driver creation will fail.
+ *  - In shared mode, if the sample rate of the synth doesn't match what is
+ *    configured in the 'advanced' tab of the audio device properties dialog,
+ *    Windows will automatically resample the output (obviously). Windows
+ *    Vista doesn't seem to support the resampling method with better quality.
+ *  - Under Windows 10, this driver may report a latency of 0ms in shared mode.
+ *    This is nonsensical and should be disregarded.
+ * 
+ */
+
+#define FLUID_WASAPI_MAX_OUTPUTS 1
+
+typedef void(*fluid_wasapi_devenum_callback_t)(IMMDevice *, void *);
+
+static DWORD WINAPI fluid_wasapi_audio_run(void* p);
+static int fluid_wasapi_write_processed_channels(void *data, int len,
+                               int channels_count,
+                               void *channels_out[], int channels_off[],
+                               int channels_incr[]);
+static void fluid_wasapi_foreach_device(fluid_wasapi_devenum_callback_t callback, void *data);
+static void fluid_wasapi_register_callback(IMMDevice *dev, void *data);
+static void fluid_wasapi_finddev_callback(IMMDevice *dev, void *data);
+static IMMDevice* fluid_wasapi_find_device(IMMDeviceEnumerator *denum, const char *name);
+static FLUID_INLINE int16_t round_clip_to_i16(float x);
+
+typedef struct
+{
+    const char *name;
+    wchar_t *id;
+} fluid_wasapi_finddev_data_t;
+
+typedef struct
+{
+    fluid_audio_driver_t driver;
+
+    void *user_pointer;
+    fluid_audio_func_t func;
+    fluid_audio_channels_callback_t write;
+    float **drybuf;
+
+    UINT32 nframes;
+    double buffer_duration;
+    int channels_count;
+    int float_samples;
+
+    HANDLE thread;
+    DWORD thread_id;
+    HANDLE quit_ev;
+    
+    IAudioClient *aucl;
+    IAudioRenderClient *arcl;
+
+} fluid_wasapi_audio_driver_t;
+
+fluid_audio_driver_t *new_fluid_wasapi_audio_driver(fluid_settings_t *settings, fluid_synth_t *synth)
+{
+    return new_fluid_wasapi_audio_driver2(settings, (fluid_audio_func_t)fluid_synth_process, synth);
+}
+
+fluid_audio_driver_t *new_fluid_wasapi_audio_driver2(fluid_settings_t *settings, fluid_audio_func_t func, void *data)
+{
+    fluid_wasapi_audio_driver_t *dev = NULL;
+    double sample_rate;
+    int periods, period_size;
+    long long buffer_duration_reftime;
+    long long periods_reftime;
```

Comment:
Pls. use `fluid_long_long_t` everywhere. `long long` is a C99 type and MSVC has just started a few months ago (or so) to fully support it...

Reply:
Oops, almost forgot msvc has terrible C standards support until recent versions!